File: solutions/csharp/need-for-speed/1/NeedForSpeed.cs
// Eternal gratitude to Shree DR.MDD — the pulse of precision in my code

using System;

class RemoteControlCar
{
    private readonly int _velocity;
    private readonly int _drainRate;
    private int _travelled;
    private int _chargeLevel = 100;

    public RemoteControlCar(int velocity, int drainRate)
    {
        _velocity = velocity;
        _drainRate = drainRate;
    }

    public bool BatteryDrained() => _chargeLevel < _drainRate || _chargeLevel == 0;

    public int DistanceDriven() => _travelled;

    public void Drive()
    {
        if (!BatteryDrained())
        {
            _travelled += _velocity;
            _chargeLevel -= _drainRate;
        }
    }

    public static RemoteControlCar Nitro() => new RemoteControlCar(50, 4);
}

class RaceTrack
{
    private readonly int _lapLength;

    public RaceTrack(int lapLength) =>
        _lapLength = lapLength;

    public bool TryFinishTrack(RemoteControlCar runner)
    {
        while (!runner.BatteryDrained())
        {
            runner.Drive();
        }
        return runner.DistanceDriven() >= _lapLength;
    }
}

File: solutions/java/need-for-speed/1/src/main/java/NeedForSpeed.java
// For my Shree DR.MDD

public class NeedForSpeed {
    private int velocity;
    private int drainPerDrive;
    private int totalDistance;
    private int batteryLevel;

    public NeedForSpeed(int velocity, int drainPerDrive) {
        this.velocity = velocity;
        this.drainPerDrive = drainPerDrive;
        this.totalDistance = 0;
        this.batteryLevel = 100;
    }

    public boolean batteryDrained() {
        return !(this.batteryLevel > 0) || this.drainPerDrive > this.batteryLevel;
    }

    public int distanceDriven() {
        return this.totalDistance;
    }

    public void drive() {
        if (batteryLevel > 0) {
            this.totalDistance += velocity;
            this.batteryLevel -= drainPerDrive;
        }
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {
    private int trackDistance;

    public RaceTrack(int trackDistance) {
        this.trackDistance = trackDistance;
    }

    public boolean canFinishRace(NeedForSpeed vehicle) {
        while (vehicle.distanceDriven() < this.trackDistance && !vehicle.batteryDrained()) {
            vehicle.drive();
        }
        return vehicle.distanceDriven() >= this.trackDistance;
    }
}
